### PR TITLE
Use different key-pair for fluentd metrics endpoint

### PR DESCRIPTION
**Pull request by josefkarasek**
Fixes BZ 1683359

### CODE REVIEW

File: fluentd/configs.d/openshift/input-pre-prometheus-metrics.conf [hunk 1]
```diff
@@ -2,9 +2,8 @@
   @type prometheus
   <ssl>
     enable true
-    certificate_path "#{ENV['ES_CLIENT_CERT']}"
-    private_key_path "#{ENV['ES_CLIENT_KEY']}"
-    ca_path "#{ENV['ES_CA']}"
+    certificate_path "#{ENV['METRICS_CERT']}"
```
Comment (richm) [inline fluentd/configs.d/openshift/input-pre-prometheus-metrics.conf]: add a default value e.g. 
```
certificate_path "#{ENV['METRICS_CERT'] || '/etc/fluent/metrics/cert'}"
```

**Comment by richm** [comment]
is there a bz for this?
do we need to backport this to 3.x?

**Comment by richm** [comment]
got a real failure - the check-zzz-fluentd-prometheus-scrape.sh test failed: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_origin-aggregated-logging/1565/pull-ci-openshift-origin-aggregated-logging-master-full-integ-aws/315/build-log.txt
```
[INFO] Logging test suite check-zzz-fluentd-prometheus-scrape started at Thu Mar 21 19:42:37 UTC 2019
Running hack/testing/check-zzz-fluentd-prometheus-scrape.sh:23: executing 'curl -k https://:24231/metrics' expecting success; re-trying every 0.2s until completion or 60.000s...
FAILURE after 59.240s: hack/testing/check-zzz-fluentd-prometheus-scrape.sh:23: executing 'curl -k https://:24231/metrics' expecting success; re-trying every 0.2s until completion or 60.000s: the command timed out
Standard output from the command:
Standard error from the command:
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: ; Unknown error
... repeated 262 times
[ERROR] PID 4945: hack/lib/cmd.sh:531: `return "${return_code}"` exited with status 1.
[INFO] 		Stack Trace: 
[INFO] 		  1: hack/lib/cmd.sh:531: `return "${return_code}"`
[INFO] 		  2: hack/testing/check-zzz-fluentd-prometheus-scrape.sh:23: os::cmd::try_until_success
[INFO]   Exiting with code 1.
[WARNING] Logging test suite check-zzz-fluentd-prometheus-scrape failed at Thu Mar 21 19:43:39 UTC 2019
```

**Comment by nhosoi** [comment]
> curl: (6) Could not resolve host: ; Unknown error

I think this is not a real failure.  Due to the failed login with -u admin -p admin, getting ${fpod} and ${fpod_ip} failed and they are empty, which causes this curl error.  I believe once the login is successful, the test would pass.  More over, in my testing with CLO pr/112 on aws does not work without this patch. ;) I wish this pr is merged asap...

**Comment by richm** [comment]
> > curl: (6) Could not resolve host: ; Unknown error
> 
> I think this is not a real failure. Due to the failed login with -u admin -p admin, getting ${fpod} and ${fpod_ip} failed and they are empty, which causes this curl error. I believe once the login is successful, the test would pass.

Other tests do not get this result.  Unless something else has changed in the meantime.  For example, a recent PR passes this test: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_origin-aggregated-logging/1563/pull-ci-openshift-origin-aggregated-logging-master-full-integ-aws/305/artifacts/full-integ-aws/logging-test-output
```[INFO] Logging test suite check-zzz-fluentd-prometheus-scrape started at Thu Mar 21 00:28:26 UTC 2019
Running hack/testing/check-zzz-fluentd-prometheus-scrape.sh:23: executing 'curl -k https://10.129.2.14:24231/metrics' expecting success; re-trying every 0.2s until completion or 60.000s...
SUCCESS after 0.160s: hack/testing/check-zzz-fluentd-prometheus-scrape.sh:23: executing 'curl -k https://10.129.2.14:24231/metrics' expecting success; re-trying every 0.2s until completion or 60.000s
[INFO] Logging test suite check-zzz-fluentd-prometheus-scrape succeeded at Thu Mar 21 00:28:27 UTC 2019
```
You can see that it has the IP address in the URL.  I don't know if it is something about this PR that breaks the test, or if something has changed in the elasticsearch-operator or cluster-logging-operator in the meantime . . .

**Comment by nhosoi** [comment]
Ah, you are right, @richm.  There's no '/etc/fluent/metrics/tls.{key,crt}' (nor METRICS_{KEY,CERT} are set) in the CI world...  In CLO, they are mounted in collection.go... :(
pkg/k8shandler/collection.go:		{Name: metricsVolumeName, MountPath: "/etc/fluent/metrics"},

I guess we should set METRICS_{KEY,CERT} to $ES_CLIENT_{KEY,CERT}, respectively in the CI (the same as before)?  But we should also consider ca_path?  Do you think this change need to be made for all the test cases not just in check-zzz-prometheus???  If that's the case, we should disable the plugin in the most of the CI test and just enable it in check-zzz-prometheus? :(

**Comment by richm** [comment]
> In CLO, they are mounted in collection.go... :(

The origin-aggregated-logging test uses whatever is the latest CLO and EO in quay.io - is there a PR for CLO that needs to be merged?

**Comment by nhosoi** [comment]
Oh, ok, then!  That’s pr/112 for CLO.  I think it’s ready to merge (although I have an alert rule file to replace, it’s small and can be in a separated pr.).

**Comment by richm** [comment]
removing autoretest until https://github.com/openshift/cluster-logging-operator/pull/112 is merged

**Comment by richm** [comment]
still missing something?

https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_origin-aggregated-logging/1565/pull-ci-openshift-origin-aggregated-logging-master-full-integ-aws/395/artifacts/full-integ-aws/scripts/check-EFK-running/artifacts/fluentd-kf62t.fluentd.log
```
2019-04-02 21:13:18 +0000 [error]: unexpected error error_class=Errno::ENOENT error="No such file or directory @ rb_sysopen - /etc/fluent/metrics/tls.crt"
  2019-04-02 21:13:18 +0000 [error]: /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/fluent-plugin-prometheus-1.3.0/lib/fluent/plugin/in_prometheus.rb:68:in `read'
  2019-04-02 21:13:18 +0000 [error]: /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/fluent-plugin-prometheus-1.3.0/lib/fluent/plugin/in_prometheus.rb:68:in `start'
...
```

**Comment by nhosoi** [comment]
> still missing something?

I wonder when CLO pr139 is picked up...  Is it immediately once it's merged?

**Comment by richm** [comment]
> > still missing something?
> 
> I wonder when CLO pr139 is picked up... Is it immediately once it's merged?

No.  It has to be built in api.ci and pushed to quay.io https://quay.io/repository/openshift/origin-cluster-logging-operator?tab=tags

**Comment by nhosoi** [comment]
> > > still missing something?
> > 
> > 
> > I wonder when CLO pr139 is picked up... Is it immediately once it's merged?
> 
> No. It has to be built in api.ci and pushed to quay.io https://quay.io/repository/openshift/origin-cluster-logging-operator?tab=tags

Does it automatically happen if we wait long enough?  Commit ID io.openshift.build.commit.id = ddf883fb11627dacb4f3934169473afedba8d331 in the quay.io repository does not include pr #139 yet...

**Comment by richm** [comment]
> Does it automatically happen if we wait long enough?

I think so - not sure exactly how it works, but it seems to show up eventually

**Comment by nhosoi** [comment]
Yay!  check-zzz-fluentd-prometheus-scrape has passed!!
```
[INFO] Logging test suite check-zzz-fluentd-prometheus-scrape started at Wed Apr  3 02:31:23 UTC 2019
Running hack/testing/check-zzz-fluentd-prometheus-scrape.sh:23: executing 'curl -k https://10.128.2.12:24231/metrics' expecting success; re-trying every 0.2s until completion or 60.000s...
SUCCESS after 0.215s: hack/testing/check-zzz-fluentd-prometheus-scrape.sh:23: executing 'curl -k https://10.128.2.12:24231/metrics' expecting success; re-trying every 0.2s until completion or 60.000s
[INFO] Logging test suite check-zzz-fluentd-prometheus-scrape succeeded at Wed Apr  3 02:31:25 UTC 2019
```

**Comment by josefkarasek** [comment]
> Yay! check-zzz-fluentd-prometheus-scrape has passed!!
> 
> ```
> [INFO] Logging test suite check-zzz-fluentd-prometheus-scrape started at Wed Apr  3 02:31:23 UTC 2019
> Running hack/testing/check-zzz-fluentd-prometheus-scrape.sh:23: executing 'curl -k https://10.128.2.12:24231/metrics' expecting success; re-trying every 0.2s until completion or 60.000s...
> SUCCESS after 0.215s: hack/testing/check-zzz-fluentd-prometheus-scrape.sh:23: executing 'curl -k https://10.128.2.12:24231/metrics' expecting success; re-trying every 0.2s until completion or 60.000s
> [INFO] Logging test suite check-zzz-fluentd-prometheus-scrape succeeded at Wed Apr  3 02:31:25 UTC 2019
> ```

How do you test it? Ideally we'd have a test case where we check from prometheus pod or we talk to prometheus programatically. The first one is simple:
```
oc exec -n openshift-monitoring prometheus-k8s-0 -- \
  curl --cacert /etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt  \
    --header "Authorization: Bearer `cat /var/run/secrets/kubernetes.io/serviceaccount/token`" \
    https://fluentd.openshift-logging.svc.cluster.local:24231/metrics
```
The second one...not sure if we can go through the oauth proxy and hit prometheus endpoint programatically.

**Comment by richm** [comment]
> through the oauth proxy

fluentd has an oauth proxy ?

**Comment by josefkarasek** [comment]
> > through the oauth proxy
> 
> fluentd has an oauth proxy ?

No, fluentd doesnt have an oauth proxy. Or a proxy of any kind..

Prometheus server has an oauth proxy. E2E test would be either
* access prometheus/targets API and check whether prometheus is successfully scaping fluentd metrics
* use selenium and do the above in a browser

The issue is that fluentd serving /metrics endpoint is just part of the story and we would ideally have e2e test that checks the entire system

**Comment by richm** [comment]
> The issue is that fluentd serving /metrics endpoint is just part of the story and we would ideally have e2e test that checks the entire system

yes - would be nice to check elasticsearch metrics in prometheus in this manner - please create a Jira card for that

**Comment by nhosoi** [comment]
Hi @josefkarasek, 
> How do you test it?

The check-zzz-fluentd-prometheus-scrape result is from ci/prow/full-integ-aws.

> oc exec -n openshift-monitoring prometheus-k8s-0 -- \
>   curl --cacert /etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt  \
>     --header "Authorization: Bearer `cat /var/run/secrets/kubernetes.io/serviceaccount/token`" \
>     https://fluentd.openshift-logging.svc.cluster.local:24231/metrics

It works just fine.  (Note, it's in my test env.)
```
$ oc exec -n openshift-monitoring prometheus-k8s-0 -- curl --cacert /etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt --header "Authorization: Bearer \$( cat /var/run/secrets/kubernetes.io/serviceaccount/token )" https://fluentd.openshift-logging.svc.cluster.local:24231/metrics | egrep -v "^#|0.0$"
fluentd_status_buffer_total_bytes{hostname="fluentd-vjxhk",plugin_id="elasticsearch-apps",plugin_category="output",type="elasticsearch"} 9869.0
....
```
But it works even w/o token...  Do we care?
```
$ oc exec -n openshift-monitoring prometheus-k8s-0 -- curl --cacert /etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt https://fluentd.openshift-logging.svc.cluster.local:24231/metrics | egrep -v "^#|0.0$"
fluentd_output_status_emit_count{hostname="fluentd-hhp8q",plugin_id="object:3fcb99e91f98",type="relabel"} 108.0
......
```

**Comment by richm** [comment]
We do not check authentication or authorization on the fluentd prometheus scrape endpoint.  We aren't exposing any sensitive information at that endpoint.  So bearer auth token headers are ignored.

I think what @josefkarasek is suggesting is that we do a test with the full metrics/prometheus stack, and have a user actually run Grafana and see the fluentd stats in Grafana.  I think we should do that, also for the Elasticsearch stats.  But that's for another PR . . .

**Comment by nhosoi** [comment]
Hi @richm, 

stop_fluentd in hack/testing/util.sh labels node like this to stop fluentd.  But it does not work for me either in my test 4.1 env...  Labeling is successful [1], but the fluentd pod on the node keeps running...  Please note that the nodes are not labeled with logging-infra-fluentd, by default [2].  Should we take some other way to stop fluentd?
`oc label node -l logging-infra-fluentd=true --overwrite logging-infra-fluentd=false`
[1]
```
$ oc describe node ip-10-0-165-98.ec2.internal | egrep logging-infra-fluentd
                    logging-infra-fluentd=false
```
[2]
```
$ oc describe node ip-10-0-167-193.ec2.internal | egrep logging-infra-fluentd
$
```
This is the failure message from zzz-rsyslog.sh in ci/prow/full-integ-aws.
```
[2019-04-04T19:16:19.264+0000] node/ip-10-0-155-19.ec2.internal labeled
[2019-04-04T19:16:19.264+0000] Running hack/testing/util.sh:666: executing 'oc get daemonset/fluentd -o jsonpath='{ .status.numberReady }'' expecting any result and text '0'; re-trying every 0.2s until completion or 120.000s...
[2019-04-04T19:16:19.264+0000] FAILURE after 120.502s: hack/testing/util.sh:666: executing 'oc get daemonset/fluentd -o jsonpath='{ .status.numberReady }'' expecting any result and text '0'; re-trying every 0.2s until completion or 120.000s: the command timed out
```

**Pull request merged**